### PR TITLE
Np 9073 unified

### PR DESCRIPTION
**Pull request by svenngjeto**
Updated karatetests and testusers creation with new login setup

### CODE REVIEW

File: cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature [hunk 1]
```diff
@@ -4,17 +4,19 @@ Feature: API tests for Cristin Person Update
...
     * def cognitoUserpoolId = java.lang.System.getenv('COGNITO_USER_POOL_ID')
     * def tokenGenerator = Java.type('no.unit.nva.cognito.CognitoUtil')
-    * def token = tokenGenerator.loginUser(username, password, cognitoUserpoolId, cognitoClientAppId)
-    * def simpleUserToken = tokenGenerator.loginUser(simple_user_name, simple_user_password, cognitoUserpoolId, cognitoClientAppId)
+    * def token = tokenGenerator.loginUser(username, password, cognitoClientAppId)
+    * def simpleUserToken = tokenGenerator.loginUser(simple_user_name, simple_user_password, cognitoClientAppId)
     * def invalidToken = 'just-a-invalid-token-for-now'
-    * def personIdentifier = '515114'
+    * def personIdentifier = '1135903'
+#    * def personIdentifier = '515114'
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature:18]: Remove comment if you want

File: cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature [hunk 2]
```diff
@@ -39,12 +41,13 @@ Feature: API tests for Cristin Person Update
     Then status 401
     And match response.message == 'Unauthorized'
 
-  Scenario: Update returns status 403 Forbidden when not authorized
+#  Scenario: Update returns status 403 Forbidden when not authorized
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature:44]: Comment here can be removed as well

File: cristin-person/src/test/java/no/unit/nva/cristin/person/handler/userAdministrator.feature [hunk 3]
```diff
@@ -39,13 +39,12 @@ Feature: API tests for Cristin Person fetch
...
-  Scenario: Fetch returns 401 Unauthorized when token is invalid token
+  Scenario: Fetch returns 200 OK and no NationalIdentificationNumber when token is invalid
     Given path '/person/' + samplePersonIdentifier
     * header Authorization = 'Bearer and.just-a.silly-text-for-token'
     When method GET
-    Then status 401
-    And match response.title == 'Unauthorized'
-#    And match responseHeaders['WWW-Authenticate'][0] == 'WWW-Authenticate'
+    Then status 200
+    And response.NationalIdentificationNumber == '#present'
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/handler/userAdministrator.feature:47]: Does this have a value?

Comment (torarnet) [reply cristin-person/src/test/java/no/unit/nva/cristin/person/handler/userAdministrator.feature:47]: I think you need to use `And match response.xxx` and also `NationalIdentificationNumber` is a field in list `identifiers` but it also has gotten a flat field directly in the model

File: cristin-person/src/test/java/no/unit/nva/cristin/person/create/createPerson.feature [hunk 4]
```diff
@@ -56,13 +56,13 @@ Feature: API tests for Cristin Person creation
...
-    When method POST
-    Then status 400
-    And response.detail = 'NationalIdentificationNumber is not valid'
+#  Scenario: Create person returns status 400 Bad Request when user is authenticated but NIN is invalid
+#    Given path '/person/'
+#    * header Authorization = 'Bearer ' + administratorToken
+#    And request sampleSimplePerson
+#    When method POST
+#    Then status 400
+#    And response.detail = 'NationalIdentificationNumber is not valid'
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/create/createPerson.feature:65]: Why comment this test?

File: cristin-commons/src/test/java/no/unit/nva/utils/AccessUtilsTest.java [hunk 5]
```diff
@@ -31,10 +29,11 @@ void validateIdentificationNumberAccess() {
     }
 
     @Test
-    void validateIdentificationNumberAccessForNewUser() throws UnauthorizedException {
+    void validateIdentificationNumberAccessForAdminUser() {
         RequestInfo requestInfo = new RequestInfo();
-        final String token = loginTestUser();
+        final String token = loginAdminTestUser();
         assertNotNull(token);
+        System.out.println(token);
```
Comment (torarnet) [inline cristin-commons/src/test/java/no/unit/nva/utils/AccessUtilsTest.java:36]: You want to print token to console?

File: cristin-commons/src/main/java/no/unit/nva/utils/UserUtils.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    public static void createUserWithRoles(String username, String password, String nationalIdentityNumber, URI customerId, Set<String> roles) throws IOException, InterruptedException {
+
+        final String poolId = AccessUtils.getUserPoolId();
+        final String appClientClientId = AccessUtils.getTestClientAppId();
+
+        createUserWithRoles(nationalIdentityNumber, customerId, roles);
+
+        CognitoUtil.deleteUser(username, poolId);
+        CognitoUtil.adminCreateUser(username, password, nationalIdentityNumber, poolId, appClientClientId);
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/utils/UserUtils.java:37]: Should this happen here or before `createUserWithRoles` ?

File: cristin-commons/src/main/java/no/unit/nva/cognito/CognitoUtil.java [hunk 7]
```diff
@@ -1,78 +1,78 @@
...
      * @return username in cognito for user created from parameters.
      */
-    public static String createAdminUser(String nvaUsername, String password, String poolId, String clientId) {
+    public static String adminCreateUser(String username, String password, String nvaUserNin, String poolId, String clientId) {
 
         List<AttributeType> list = new ArrayList<>();
-        list.add(newAttribute(nvaUsername, NVA_USERNAME_ATTRIBUTE));
         list.add(newAttribute(CURRENT_CUSTOMER_CLAIM_VALUE.toString(), CURRENT_CUSTOMER_CLAIM_NAME));
-        list.add(newAttribute(ACCESS_RIGHTS_CLAIM_NAME, ACCESS_RIGHTS_CLAIM_VALUE));
+        list.add(newAttribute(ACCESS_RIGHTS_CLAIM_VALUE, ACCESS_RIGHTS_CLAIM_NAME));
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/cognito/CognitoUtil.java:69]: name or value first?

**Review by torarnet — COMMENTED** [review]
See comments

File: cristin-person/src/test/java/no/unit/nva/cristin/person/handler/userAdministrator.feature [hunk 8]
```diff
@@ -4,48 +4,48 @@ Feature: API tests for Cristin Person fetch
...
     * def tokenGenerator = Java.type('no.unit.nva.cognito.CognitoUtil')
-    * def token = tokenGenerator.loginUser(username, password, cognitoUserpoolId, cognitoClientAppId)
-    * def simpleUserToken = tokenGenerator.loginUser(simple_user_name, simple_user_password, cognitoUserpoolId, cognitoClientAppId)
+    * def token = tokenGenerator.loginUser(username, password, cognitoClientAppId)
+    * def simpleUserToken = tokenGenerator.loginUser(simple_user_name, simple_user_password, cognitoClientAppId)
     * def samplePersonIdentifier = '515114'
     * def PROBLEM_JSON_MEDIA_TYPE = 'application/problem+json'
     Given url CRISTIN_BASE
     * print 'Current base url: ' + CRISTIN_BASE
+    * print 'token= ' + token
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/handler/userAdministrator.feature:20]: You want to print this to console?

File: cristin-commons/src/main/java/no/unit/nva/utils/AccessUtils.java [hunk 9]
```diff
@@ -63,4 +82,45 @@ private static boolean requesterHasNoAccessRightToUseNationalIdentificationNumbe
     private static boolean requesterHasNoAccessRightToEditProjects(RequestInfo requestInfo) {
         return !requestInfo.userIsAuthorized(EDIT_OWN_INSTITUTION_PROJECTS);
     }
+
+    public static String getBackendAccessToken() throws IOException, InterruptedException {
+        var cognitoTokenUrl = URI.create("https://nva-dev.auth.eu-west-1.amazoncognito.com/oauth2/token");
```
Comment (brinxmat) [inline cristin-commons/src/main/java/no/unit/nva/utils/AccessUtils.java:87]: This URI is hardcoded to dev, eu, etc. Will this fix beyond (also outside this string) work across envs?

File: cristin-commons/src/test/java/no/unit/nva/utils/UserUtilCreateUsers.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+import static no.unit.nva.utils.AccessUtils.getTestClientAppId;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+@Tag("integrationTest")
+@Tag("createTestUsers")
+public class UserUtilCreateUsers {
+
+    private static final Logger logger = LoggerFactory.getLogger(UserUtilCreateUsers.class);
+    public static final String CREATE_TEST_USER_WITH_USER_NAME = "createTestUser with UserName {}";
+    public static final URI CUSTOMER = URI.create("https://api.dev.nva.aws.unit.no/customer/bb3d0c0c-5065-4623-9b98-5810983c2478");
```
Comment (brinxmat) [inline cristin-commons/src/test/java/no/unit/nva/utils/UserUtilCreateUsers.java:29]: dev…

File: cristin-commons/src/test/java/no/unit/nva/utils/UserUtilsTest.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+@Tag("integrationTest")
+class UserUtilsTest {
+
+    @Test
+    void createUserWithRoles() throws IOException, InterruptedException {
+
+        String username = "ninja@test";
+        String password = randomString();
+        String nin = "27034425812";
```
Comment (brinxmat) [inline cristin-commons/src/test/java/no/unit/nva/utils/UserUtilsTest.java:20]: dev…

File: docs/cristin-proxy-swagger.yaml [hunk 12]
```diff
@@ -666,6 +666,8 @@ paths:
       summary: Find person by identifier
       description: Returns a single person matching id parameter
       operationId: GetPersonById
+#      security:
```
Comment (brinxmat) [inline docs/cristin-proxy-swagger.yaml:669]: ?

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: cristin-commons/src/main/java/no/unit/nva/cognito/CognitoUtil.java (hunk 7 above)
Comment (svenngjeto) [reply cristin-commons/src/main/java/no/unit/nva/cognito/CognitoUtil.java:69]: Swapped parameters into more natural order

File: docs/cristin-proxy-swagger.yaml (hunk 12 above)
Comment (svenngjeto) [reply docs/cristin-proxy-swagger.yaml]: Endpoint is currently  without authorization and without NIN in response

File: cristin-commons/src/test/java/no/unit/nva/utils/UserUtilsTest.java (hunk 11 above)
Comment (svenngjeto) [reply cristin-commons/src/test/java/no/unit/nva/utils/UserUtilsTest.java:20]: It's a integration test for manually checking if environment is correct set up,  will not not as part of deployment

File: cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature [hunk 13]
```diff
@@ -39,12 +40,12 @@ Feature: API tests for Cristin Person Update
     Then status 401
     And match response.message == 'Unauthorized'
 
-  Scenario: Update returns status 403 Forbidden when not authorized
+  Scenario: Update returns status 401 Unauthorized when not authorized because of getPersonCristinId
```
Comment (torarnet) [inline cristin-person/src/test/java/no/unit/nva/cristin/person/update/cristinPersonUpdate.feature:43]: Add a karate test where you mock getPersonCristinId using HandlerRequestBuilder and see if you get a forbidden after that when you dont have required access rights. Maybe in separate PR if you want to get this done.

File: cristin-commons/src/main/java/no/unit/nva/utils/AccessUtils.java [hunk 14]
```diff
@@ -63,4 +92,52 @@ private static boolean requesterHasNoAccessRightToUseNationalIdentificationNumbe
...
+                .header(CONTENT_TYPE, APPLICATION_X_WWW_FORM_URLENCODED)
+                .header(AUTHORIZATION, basicAuthHeader())
+                .build();
+        HttpClient client = HttpClient.newBuilder()
+                .followRedirects(HttpClient.Redirect.ALWAYS)
+                .connectTimeout(Duration.ofSeconds(30))
+                .build();
+
+        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
+        logger.debug("Response from Cognito: statusCode={}, body:{}", response.statusCode(), response.body());
```
Comment (torarnet) [inline cristin-commons/src/main/java/no/unit/nva/utils/AccessUtils.java:112]: Does this print token as it prints body? Maybe not important when using debug?

**Review by torarnet — APPROVED** [review]
Approved but please get more approvals because of a lot of authentication stuff

**Pull request merged**